apollo-federation/src/sources/connect/json_selection/README.md

```diff
@@ -574,25 +637,77 @@ registered in the `JSONSelection` parser in order to be recognized.
 list may grow and/or become user-configurable in the future:
 
-> Full disclosure: even this list is still aspirational, but suggestive of the
-> kinds of methods that are likely to be supported in the next version of the
-> `JSONSelection` parser.
-
 ```graphql
-list->first { id name }
-list->last.name
-list->slice($args.start, $args.end)
-list->reverse
-some.value->times(2)
-some.value->plus($addend)
-some.value->minus(100)
-some.value->div($divisor)
-isDog: kind->eq("dog")
-isNotCat: kind->neq("cat")
-__typename: kind->match({ "dog": "Dog", "cat": "Cat" })
-decoded: utf8Bytes->decode("utf-8")
-utf8Bytes: string->encode("utf-8")
-encoded: bytes->encode("base64")
+# The ->echo method returns its first input argument as-is, ignoring
+# the input data. Useful for embedding literal values, as in
+# $->echo("give me this string"), or wrapping the input value.
+__typename: $->echo("Book")
+wrapped: field->echo({ fieldValue: @ })
+
+# Returns the type of the data as a string, e.g. "object", "array",
+# "string", "number", "boolean", or "null". Note that `typeof null` is
+# "object" in JavaScript but "null" for our purposes.
+typeOfValue: value->typeof
+
+# When invoked against an array, ->map evaluates its first argument
+# against each element of the array, binding the element values to `@`,
+# and returns an array of the results. When invoked against a non-array,
+# ->map evaluates its first argument against that value and returns the
+# result without wrapping it in an array.
+doubled: numbers->map(@->mul(2))
+types: values->map(@->typeof)
+
+# Returns true if the data is deeply equal to the first argument, false
+# otherwise. Equality is solely value-based (all JSON), no references.
+isObject: value->typeof->eq("object")
+
+# Takes any number of pairs [candidate, value], and returns value for
+# the first candidate that equals the input data. If none of the
+# pairs match, a runtime error is reported, but a single-element
+# [<default>] array as the final argument guarantees a default value.
+__typename: kind->match(
+    ["dog", "Canine"],
+    ["cat", "Feline"],
+    ["Exotic"]
+)
+
+# Like ->match, but expects the first element of each pair to evaluate
+# to a boolean, returning the second element of the first pair whose
+# first element is true. This makes providing a final catch-all case
+# easy, since the last pair can be [true, <default>].
+__typename: kind->matchIf(
+    [@->eq("dog"), "Canine"],
+    [@->eq("cat"), "Feline"],
+    [true, "Exotic"]
+)
+
+# Arithmetic methods, supporting both integers and floating point values,
+# similar to JavaScript.
+sum: $.a->add($.b)->add($.c)
+difference: $.a->sub($.b)->sub($.c)
+product: $.a->mul($.b, $.c)
+quotient: $.a->div($.b)
+remainder: $.a->mod($.b)
+
+# Array/string methods
+first: list->first
+last: list->last
+index3: list->get(3)
+secondToLast: list->get(-2)
+# The ->get method also works for object keys.
+aValue: $->echo({ a: 123 })->get("a")
+slice: list->slice(0, 5)
```

Comment:
Does this also work with `-` indexing? Does slicing past the end of an array generate a runtime error, return the size of the array filled with `null`, truncate, or something else?

Reply:
Yes, negative indexing works as an offset from the end of the array.

Slicing past the end of the array works like JavaScript, where the resulting array won't contain more elements than were available from the original array, so you don't get any `null` placeholder values.

These behaviors are all well-defined and enforced with runtime errors in the [`slice_method`](https://github.com/apollographql/router/blob/89e6022f0b35aeef32ccf53fdb570f28bc0eafe9/apollo-federation/src/sources/connect/json_selection/methods.rs#L688) implementation. In Rust terms, the past-the-end behavior comes down to the behavior of the `.take(end - start)` method, whose comment says:
```rs
    /// Creates an iterator that yields the first `n` elements, or fewer
    /// if the underlying iterator ends sooner.
```